### PR TITLE
Add symlink support for readDirectory

### PR DESCRIPTION
**Pull request by km-tr**
### Description

**Issue**  
Since **v3.13.0**, the **Cline Rules pop‑over** does not load rule files delivered through **symbolic links**.  
`readDirectory()` was skipping entries where `Dirent.isSymbolicLink()` is `true`, so the pop‑over could not discover those rules.

**Fix**  
* Update `src/utils/fs.ts` so that `readDirectory()`  
  * Detects symlinks, resolves them with `fs.readlink()`, and inspects the target stats.  
  * Returns symlinked files and recursively traverses symlinked directories.  
* Add comprehensive unit tests in `src/utils/fs.test.ts`.  
* Add a changeset describing the bug‑fix.

### Test Procedure

1. `npm test` — new test cases cover:  
   * Standard directory traversal.  
   * Symlink → file.  
   * Symlink → directory (recursive).  
   * Graceful failure on missing directories.  
2. Manual QA: run Cline with a rules folder containing symlinks and confirm the pop‑over lists them.  
3. Regression: verify circular symlinks are safely ignored and performance remains unaffected.

### Type of Change

-   [x] 🐛 Bug fix (non-breaking change which fixes an issue)
-   [ ] ✨ New feature (non-breaking change which adds functionality)
-   [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] ♻️ Refactor Changes
-   [ ] 💅 Cosmetic Changes
-   [ ] 📚 Documentation update
-   [ ] 🏃 Workflow Changes

### Pre-flight Checklist

-   [x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
-   [x] Tests are passing (`npm test`) and code is formatted and linted (`npm run format && npm run lint`)
-   [x] I have created a changeset using `npm run changeset` (required for user-facing changes)
-   [x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

### Screenshots

_N/A_

### Additional Notes

* **Performance impact:** minimal — traversal is still depth‑first and filters OS artefacts before symlink resolution.  
* **Edge cases:** circular symlinks are not followed, preventing infinite loops.

----

> [!IMPORTANT]
> Add symlink support to `readDirectory()` in `fs.ts`, with tests in `fs.test.ts`.
> 
>   - **Behavior**:
>     - `readDirectory()` in `fs.ts` now resolves symlinks using `fs.readlink()` and inspects target stats.
>     - Supports symlinks to files and directories, recursively traversing symlinked directories.
>     - Safely ignores circular symlinks to prevent infinite loops.
>   - **Tests**:
>     - Added unit tests in `fs.test.ts` for symlink handling, including symlink to file, symlink to directory, and error handling for non-existent directories.
>   - **Misc**:
>     - Added a changeset `soft-windows-melt.md` describing the bug fix.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for 13ce4b76a7317a5f78df6d001b8d155b70a1c490. You can [customize](https://app.ellipsis.dev/cline/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: src/utils/fs.ts [hunk 1]
```diff
@@ -84,18 +84,44 @@
...
+
+		const filePaths: string[] = []
+
+		for (const entry of filteredEntries) {
+			const fullPath = path.resolve(entry.parentPath, entry.name)
+
+			if (entry.isFile()) {
+				filePaths.push(fullPath)
+			} else if (entry.isSymbolicLink()) {
+				const targetPath = await fs.readlink(fullPath)
```
[the remark this answers is not in the document]
Comment (km-tr) [reply src/utils/fs.ts:108]: Good suggestion. In this context we only need the immediate link target, not its fully‑resolved absolute path, so fs.readlink is sufficient and avoids an extra stat call. If we later need full resolution (e.g. nested links across filesystems) we can revisit this, but for now I’ll keep readlink to minimise I/O overhead.

File: src/utils/fs.ts [hunk 2]
```diff
@@ -84,18 +84,44 @@
...
+				const resolvedPath = path.resolve(path.dirname(fullPath), targetPath)
+
+				const stats = await fs.stat(resolvedPath)
```
[the remark this answers is not in the document]
Comment (km-tr) [reply src/utils/fs.ts]: The surrounding function already propagates I/O errors up to the caller, where they’re surfaced as user‑friendly messages. Adding an inner try/catch would hide those errors and make debugging harder. For that reason I prefer to let the exception bubble up unchanged.

File: src/utils/fs.ts [hunk 3]
```diff
@@ -84,18 +84,44 @@ const OS_GENERATED_FILES = [
...
+
+				if (stats.isFile()) {
+					filePaths.push(fullPath)
+				} else if (stats.isDirectory()) {
+					const nestedFiles = await readDirectory(resolvedPath)
```
[the remark this answers is not in the document]
Comment (km-tr) [reply src/utils/fs.ts]: Implemented cycle detection using a visitedPaths set to track already processed symlink targets and prevent infinite recursion. The function now skips symlinks pointing to already visited paths and
  gracefully handles broken symlinks by catching errors during resolution.

File: src/utils/fs.test.ts [hunk 4]
```diff
@@ -88,4 +88,66 @@ describe("Filesystem Utilities", () => {
...
+
+		it("should throw an error for non-existent directories", async () => {
+			const nonExistentDir = path.join(tmpDir, "does-not-exist")
+			try {
+				await readDirectory(nonExistentDir)
+			} catch (error) {
+				error.should.be.an.instanceOf(Error)
+			}
+		})
+	})
```
[the remark this answers is not in the document]
Comment (km-tr) [reply src/utils/fs.test.ts:195]: done.

**Comment by 0xToshii** [comment]
@km-tr thanks for contributing! sorry it took a while to get to this - is this a big issue you're still facing?

couple potential issues for this implementation but the main one is that the symlink handling here is only handles it if its one level deep. im really conflicted on including explicit symlink handling so did want to get your thoughts on that

**Comment by km-tr** [comment]
@0xToshii 
I do think it would be convenient to have.
As for the fact that symbolic link handling is currently limited to just one level here, that was an intentional choice to avoid added complexity for now.

**Comment by 0xToshii** [comment]
@km-tr i think this code has some edge cases where we aren't correctly handling cycles due to the file path representations of the same file when you traverse different symlinks which ultimately point to the same place. i think we need to have a broader conversation about this, and think through the optimal implementation, which ive created a feature request for here: #4445
